util/gen_doc_hw_summary_table.py

```diff
@@ -0,0 +1,193 @@
+        one_liner = data['one_line_desc']
+
+        # Create HW summary entry and append it to our list.
+        summary_entries.append(HwSummaryEntry(name, one_liner))
+
+    # Create table summarizing HW blocks.
+    table = HwSummaryTable(summary_entries)
+
+    # Write table into hw/ip/README.md, replacing any content that was
+    # previously in the section marked as autogenerated by this tool.
+    with open(hw_ip_dir / 'README.md', 'r') as f:
+        text = f.read()
+        new_section = (
+            '<!-- Do NOT edit this section manually, '
+            f'it is generated by {os.path.relpath(this_file_path, repo_root)}. '
+            '-->\n')
+        new_section += table.to_markdown().rstrip('\n')
+        new_text = replace_section_in_text(
+            text,
+            '\n<!-- BEGIN AUTOGEN Summary Table -->\n',
+            '\n<!-- END AUTOGEN Summary Table -->\n',
+            new_section)
+    with open(hw_ip_dir / 'README.md', 'w') as f:
+        f.write(new_text)
+
+
+@dataclass
+class HwSummaryEntry:
+    """Simple dataclass summarizing one HW block."""
+    block_name: str
+    one_line_desc: str
+
+    def name_as_linked_markdown(self) -> str:
+        return f'[{self.block_name}]({self.block_name})'
+
+
+@dataclass
+class HwSummaryTable:
+    """Table summarizing multiple hardware blocks, can be formatted as Markdown.
+    """
+    entries: List[HwSummaryEntry]
+
+    def to_markdown(self) -> str:
+        header = ['HW Block', 'Brief Summary']
+        data = [
+            [e.name_as_linked_markdown() for e in self.entries],
+            [e.one_line_desc for e in self.entries]
+        ]
+        assert len(header) == len(data), (
+            "number of table headers does not match number of data columns")
+
+        col_widths = []
+        for h, d in zip(header, data):
+            col_entries = [h] + d
+            col_width = max(len(c) for c in col_entries)
+            # Add 2 to width to make room for one space on the left and one on
+            # the right.
+            col_width += 2
+            col_widths.append(col_width)
+
+        def left_align(text: str, width: int) -> str:
+            num_right_spaces = width - len(text) - 1
+            assert num_right_spaces >= 0
+            right_spaces = ' ' * num_right_spaces
+            return ' ' + text + right_spaces
+
+        # Write header.
+        text = ''
+        for txt, width in zip(header, col_widths):
+            text += '|' + left_align(txt, width)
+        text += '|\n'
+
+        # Write line separating header and data rows.
+        for col_width in col_widths:
+            text += '|' + ('-' * col_width)
+        text += '|\n'
+
+        # Transpose data from a list of columns to a list of rows.
+        data = [list(r) for r in zip(*data)]
+
+        # Write data rows.
+        for row in data:
+            for txt, width in zip(row, col_widths):
+                text += '|' + left_align(txt, width)
+            text += '|\n'
```

Comment:
OT already has `tabulate` as a dependency, which can handle generating markdown tables.

https://github.com/lowRISC/opentitan/blob/ff2806c8a84c0011d9a9427e37863d37ef499877/python-requirements.txt#L17

But this code is short and nicely written, so I wouldn't object to keeping it.

Reply:
Thanks, that gives me a possibility to remove potentially unnecessary code, which is always good :-) I'll give it a try

Reply:
Done